transaction/summary/src/report.rs

```diff
@@ -3,90 +3,368 @@
+        {
+            // If we have an entry, subtract the change value from this
+            Some(v) => v.2 = v.2.checked_sub(value).ok_or(Error::NumericOverflow)?,
+            // If we do not, create a new entry
+            None => self
+                .totals
+                .push((token_id, TotalKind::Ours, -value))
+                .map_err(|_| Error::BufferOverflow)?,
+        }
+
         Ok(())
     }
 
-    /// This should be done before displaying the report
+    /// Add SCI (or other) input not owned by our account
+    fn sci_add(&mut self, amount: Amount) -> Result<(), Error> {
+        let Amount { token_id, value } = amount;
+
+        // Ensure value will not overflow
+        let value = i64::try_from(value).map_err(|_| Error::NumericOverflow)?;
+
+        // Check for existing total entry for this token
+        match self
+            .totals
+            .iter_mut()
+            .find(|(t, k, _)| t == &token_id && *k == TotalKind::Sci)
+        {
+            // If we have an entry, add the value to this
+            Some(v) => v.2 = v.2.checked_add(value).ok_or(Error::NumericOverflow)?,
+            // If we do not, create a new entry
+            None => self
+                .totals
+                .push((token_id, TotalKind::Sci, value))
+                .map_err(|_| Error::BufferOverflow)?,
+        }
+        Ok(())
+    }
+
+    /// Add output value to a particular entity / address to the report
+    fn output_add(&mut self, entity: TransactionEntity, amount: Amount) -> Result<(), Error> {
+        let Amount { token_id, value } = amount;
+
+        // Check for existing output for this address
+        match self
+            .outputs
+            .iter_mut()
+            .find(|(e, t, _)| t == &token_id && e == &entity)
+        {
+            // If we have an entry, subtract the change value from this
+            Some((_, _, v)) => *v = v.checked_add(value).ok_or(Error::NumericOverflow)?,
+            // If we do not, create a new entry
+            None => self
+                .outputs
+                .push((entity, token_id, value))
+                .map_err(|_| Error::BufferOverflow)?,
+        }
+
+        Ok(())
+    }
+
+    /// Add network fee to the report
+    fn network_fee_set(&mut self, amount: Amount) -> Result<(), Error> {
+        // Set fee value
+        self.network_fee = amount;
+
+        Ok(())
+    }
+
+    /// Set tombstone block in the report
+    fn tombstone_block_set(&mut self, value: u64) -> Result<(), Error> {
+        self.tombstone_block = value;
+        Ok(())
+    }
+
+    /// Finalise report, checking and balancing totals and sorting report
+    /// entries
+    fn finalize(&mut self) -> Result<(), Error> {
+        // Sort outputs and totals
+        self.sort();
+
+        // For each token id, check that inputs match outputs
+        // (this is only executed where _totals_ exist, so skipped
+        // for the current SCI implementation)
```

Comment:
Hm, why wouldn't `totals` exist for an SCI? `.sci_add()` updates it

Reply:
oop i think that's an old comment